lib/plugin-loader.ts

```diff
@@ -0,0 +1,853 @@
+            //`Plugin ${plugin.identifier} at path=${plugin.location} loaded.\n`);
+            bootstrapLogger.debug("ZWED0164I", plugin.toString()); //bootstrapLogger.debug(' Content:\n' + plugin.toString());
+          let frozen = zluxUtil.deepFreeze(plugin);
+          this.plugins.push(frozen);
+          newPlugins.push(frozen);
+          this.pluginMap[plugin.identifier] = plugin;
+          successCount++;
+        } else {
+          bootstrapLogger.info("ZWED0125I", pluginDef.identifier); //bootstrapLogger.log(bootstrapLogger.INFO,
+            //`Plugin ${pluginDef.identifier} not loaded`);
+          this.pluginMap[pluginDef.identifier] = {}; //mark it as having failed so it will not be retried
+        }
+      } catch (e) {
+        bootstrapLogger.warn("ZWED0035W", pluginDef.identifier, e);
+        //downstream will pick up error
+        newPlugins.push(Object.assign(pluginDef, {error:e}));
+      }
+    }
+    this.registerStaticPluginsWithManagers(sortedAndRejectedPlugins.plugins);
+    for (const plugin of newPlugins) {
+      this.emit('pluginFound', {
+        data: plugin,
+        count: newPlugins.length
+      });
+    }
+  }
+
+  public issueRefreshFinish() {
+    this.emit('refreshFinish', {});
+  }
+
+  /**
+     Language managers that need to know about all plugins in advance should be informed here.
+     These managers may or may not support dynamic addition as well.
+   */
+  public registerStaticPluginsWithManagers(pluginDefs: any) {
+    if (this.options.langManagers) {
+      for (let i = 0; i < this.options.langManagers.length; i++) {
+        this.options.langManagers[i].registerPlugins(pluginDefs);
+      }
+    }
+  }
+  
+  public addDynamicPlugin(pluginDef: any) {
+    if (this.pluginMap[pluginDef.identifier]) {
+      throw new Error('ZWED0019E - Plugin already registered');
+    }
+    const pluginContext = {
+      productCode: this.options.productCode,
+      config: this.options.serverConfig,
+      authManager: this.options.authManager
+    };
+    //
+    //TODO resolving dependencies correctly
+    //see also: the FIXME note at the end of installPlugins()
+    //
+    bootstrapLogger.info("ZWED0046I", pluginDef.identifier); //bootstrapLogger.info("Adding dynamic plugin " + pluginDef.identifier);
+    const pluginConfiguration = configService.getPluginConfiguration(
+      pluginDef.identifier, pluginDef.location,
+      this.options.serverConfig, this.options.productCode);
+    const plugin = new makePlugin(pluginDef, pluginConfiguration, null, pluginContext,
+      true);
+//    if (!this.unresolvedImports.allImportsResolved(pluginContext.plugins)) {
+//      throw new Error('unresolved dependencies');
+//      this.unresolvedImports.reset();
+//    }
+    zluxUtil.deepFreeze(plugin);
+    this.plugins.push(plugin);
+    this.pluginMap[plugin.identifier] = plugin;
+    this.emit('pluginAdded', {
+      data: plugin
+    });
+  }
+};
+
+module.exports = PluginLoader;
```

Comment:
Believe it or not, you do not need to use `module.exports` in typescript. The reason why is because the `import` and `export` system compiles into `module.exports` in the case that the compile type is nodejs. `export` would do something different if compiled for a browser. there are a few choices.